Fix unrecognized selector error for 'taskIdentifier' and 'originalRequest' methods

nebiros opened the pull request:
Adds support to integrate DVR with Alamofire. Fix: https://github.com/venmo/DVR/pull/41#issuecomment-171917800.

I made a little code snippet about it: https://coderwall.com/p/sd7i2g/my-alamofire-dvr-recipe-for-network-testing-in-swift

nebiros committed:
Fix unrecognized selector error for 'taskIdentifier' and 'originalRequest' methods

nebiros committed:
Merge pull request #1 from nebiros/fix-unrecognized-selector

Fix unrecognized selector error for 'taskIdentifier' and 'originalRequest' methods

eliperkins commented:
Hm. Is there more context around what the bug is here? I see you mentioned that there's an unrecognized selector error, but what causes it?

I'd be hesitant to move forward with merging this without knowing who/what is calling on those selectors.

nebiros commented:
@eliperkins ouch, sorry, fix this: https://github.com/venmo/DVR/pull/41#issuecomment-171917800, to integrate DVR with Alamofire, I made a little code snippet about it: https://coderwall.com/p/sd7i2g/my-alamofire-dvr-recipe-for-network-testing-in-swift

nebiros commented:
@eliperkins any update?

eliperkins commented:
I'd love to get other folks eyes on this before merging it, as I don't maintain an up-to-date DVR test suite anymore.

@soffes @hyperspacemark or other contributors, would you mind reviewing this as well?

nebiros committed:
Merge branch 'master' of github.com:nebiros/DVR

dmiluski commented:
Hi @nebiros , could we update this PR with a description of the problem, the expectation of performance, then how the fix tackles that?

nebiros commented:
@dmiluski sure, I update the description

nebiros committed:
Fixes unrecognized selector for 'taskIdentifier' and 'originalRequest' methods

- Adds 'taskIdentifier' and 'originalRequest' methods for SessionUploadTask
and SessionDownloadTask

nebiros committed:
Merge remote-tracking branch 'upstream/master'

nebiros commented:
@dmiluski @eliperkins ping

eliperkins reviewed (changes requested):
I don't think this is the correct approach to take to implement this.

According to the [`URLSessionTask` docs](https://developer.apple.com/documentation/foundation/nsurlsessiontask/1411231-taskidentifier):

> This value is unique only within the context of a single session; tasks in other sessions may have the same `taskIdentifier` value.

Taking a look at how this is implemented in [`swift-corelibs-foundation`](https://github.com/apple/swift-corelibs-foundation/blob/eff149c5fb5d38f92dcca4fa2f01ec3ca5255571/Foundation/URLSession/URLSession.swift#L424-L425), it looks like the approach taken is to allow the **session** to hold on to the task identifier counter, and to dependency inject it to each task, using an `internal` initializer.

To me, this is the more correct approach, and one that would not allow for task identifiers to clash, should you reach a sufficient amount of them.

nebiros committed:
Merge remote-tracking branch 'upstream/master'

nebiros committed:
Merge remote-tracking branch 'upstream/master'
